src/server/tran_server.cpp

```diff
@@ -552,33 +552,32 @@ tran_server::connection_handler::disconnect_async (bool with_disc_msg)
   m_disconn_future = std::async (std::launch::async, [this, with_disc_msg]
   {
     on_disconnecting (); // server-type specific jobs before disconnecting.
-    // m_conn is not nullptr since it's only set to nullptr here below once
 
-    m_conn->stop_response_broker (); // wake up threads waiting for a response and tell them it won't be served.
-    auto ulock_conn = std::unique_lock<std::shared_mutex> { m_conn_mtx };
+    /*
+     * Stop incoming communication and wake up threads waiting for a response and tell them it won't be served.
+     * This has to be out of the lock of m_conn_mtx because waitors are holding the lock. (send_receive())
+     * m_conn is not nullptr here since it's only set to nullptr here below.
+     */
+    m_conn->stop_incoming_communication_thread ();
+
+    auto lockg_state = std::lock_guard { m_state_mtx };
+    auto lockg_conn = std::lock_guard { m_conn_mtx };
     const std::string channel_id = get_channel_id ();
+
+    assert (m_state == state::DISCONNECTING);
+
     if (with_disc_msg)
       {
 	const int payload = static_cast<int> (m_ts.m_conn_type);
 	std::string msg (reinterpret_cast<const char *> (&payload), sizeof (payload));
 	m_conn->push (tran_to_page_request::SEND_DISCONNECT_MSG, std::move (msg));
-	// After sending SEND_DISCONNECT_MSG, the page server may release all resources releated to this connection.
+	// After sending SEND_DISCONNECT_MSG, the page server may release all resources related to this connection.
+	// So, it has to be the last msg.
       }
 
-    // stop_incoming_communication_thread() has to be done explicitly before m_conn.reset () to avoid a request handler or an error handler accesses nullptr of m_conn.
-    m_conn->stop_incoming_communication_thread ();
-    m_conn->stop_outgoing_communication_thread ();
```

Comment:
> request_queue_autosend<ReqQueue>::~request_queue_autosend could be removed altogether from everywhere.

Do you happen to mean `stop_outgoing_communication_thread`? Or do you mean we should call `stop_outgoing_communication_thread` explicitly everywhere?

Reply:
Sorry, my previous message is confusing even for me now. :(

I meant `request_sync_client_server::stop_outgoing_communication_thread` could be removed altogether from everywhere.
And then the queue autosend will be stopped via: `request_sync_client_server` dtor -> `request_queue_autosend` dtor -> `request_queue_autosend::stop_thread`.

The goal of removing it is to keep the same approach on the xTS and PS side.

Reply:
No problem. I think we can remove it and make it consistent, but I think it'd be better to postpone the refactoring later. It's the fair of the `stop_incoming_communication_thread`, and allow us a kind of "read-only" connection_handler. I don't know it will be needed. But, I think we can refactor it later when the connection_handler's are more mature and stabilized. Or, do you think it is better to refactor as many as possible we can and it's kind of "good way"? No sarcasm at all, I just want to listen to your insight.